doc/user-manual/language/type-based-termination-checking.rst

```diff
@@ -0,0 +1,317 @@
+
+::
+
+    repeat : Stream
+    repeat .head = zero
+    repeat .tail = repeat
+
+We shall again focus on the second branch ``Stream.tail = repeat``, as it is the only branch relevant from a termination perspective. Assume that ``repeat`` produces a stream of depth ``n``. According to the definition of ``tail``, this branch needs to construct a stream of depth ``m < n`` *for any* ``m``. A direct recursive call to ``repeat`` suffices here: it can be assumed that the inner ``repeat`` is used with the depth ``m``. Now, since the stream-returning function is defined in terms of "shallower" streams, Agda considers it terminating, as an arbitrary number of unfoldings for ``repeat`` will terminate.
+
+Now consider the following function:
+
+::
+
+    badRepeat : Stream
+    badRepeat .head = zero
+    badRepeat .tail = badRepeat .tail
+
+The difference here is that now inner ``badRepeat`` is projected. The logic from the previous paragraph does not apply here: if ``badRepeat .tail`` is of depth ``m``, then the inner ``badRepeat`` must have depth bigger than ``m``, say ``k``. There is no evidence that ``k < n``, so Agda rejects this definition as non-terminating. Indeed, it can be unfolded infinitely, which destroys strong normalization.
+
+.. _type-based-termination-checking-mutual-induction-coinduction:
+
+Induction-Coinduction
+-----------------------
+
+One feature of the type-based termination checker is its support for mutually defined inductive and coinductive types.
+
+We will illustrate this using stream processors.
+
+::
+
+    -- The inductive part of stream processors
+    -- This datatype allows choosing to consume incoming stream or
+    -- produce an element of a new one.
+    data   SPμ (A B : Set) : Set
+    -- The coinductive part of stream processors
+    -- This datatype carries infinite behavior of this class.
+    record SPν (A B : Set) : Set
+
+    data SPμ A B where
+      get : (A → SPμ A B) → SPμ A B -- Consumes an element of the incoming stream
+      put : B → SPν A B → SPμ A B   -- Produces an element of the constructed stream
+
+    record SPν A B where
+      coinductive
+      field force : SPμ A B
+
+    open SPν
+
+This datatype can be understood differently depending on the order of fixpoint operators in its formal definition. The two meanings that can be given here are:``ν Y. μ X. (A → X) + (B * Y)`` and ``μ X. ν Y. (A → X) + (B * Y)``.There is a substantial difference here. The first interpretation means that the stream processor infinitely produces ``B``, consuming a finite number of ``A``\s between two productions. The second interpretation means that the stream processor may consume only a finite number of ``A``\s, and between each two consumptions, it is allowed to produce an infinite number of ``B``\s. It is natural to select the first interpretation for stream processors, and that's what the type-based termination checker does.
+
+This behavior can be reflected in the use of Agda's sized types:
```

Comment:
This doc page is not about sized types, so I would avoid using them here for an example. Instead, just explain how the type-based termination checker treats the example in the absence of explicit size annotations.